Add support for device provisioning

aebruno opened the pull request:
First thanks for the excellent library. This pull request is a first cut at supporting device linking. It's based off the code in libsignal-service-java. Added support for fetching linked devices, linking a new device, and removing a device. I did some initial testing using Signal-Desktop in a development environment and seems to be working (aside from contact syncing). I  added an example of how the client API would work into `cmd/textsecure/main.go` command. Basically, you fire up Signal-Desktop scan the QR code and pass in the URL:

    $ textsecure -linkdevice 'tsdevice:/?uuid=xxx&pub_key=yyy'

If all goes well, to see linked devices you can run:

    $ ./textsecure -showdevices

Currently, it looks like Open Whisper Systems only supports device linking for Android, so it may be worth contacting them at some point to ensure this wouldn't be in violation of their terms, etc.

That all being said, I am in no way a crypto expert but hope this PR can at least serve as starting point. Feedback welcome!

Thanks again for the awesome work on this library.

janimo commented:
Thanks, this looks good! I'll try reviewing and merging this soon.
AFAIK OWS do not have iOS devices syncing because they lacked the resources for iOS development.

aebruno commented:
Sounds great. Let me know if you'd like anything changed. Contact syncing is the other piece to this and looks like it's done via SyncMessages. I'm working on fleshing this out as well and will perhaps submit a separate PR for that.

aebruno commented:
@janimo I added support for SyncMessages to this PR which was the final piece to getting device linking working. Contact/Group syncing is now working as well as the various Sync message types. I tested this with Signal-Desktop and things seem to be working nicely. 

Tried my best not to make these changes too invasive. No changes were made to the public facing client API. I had to refactor buildAndSendMessages slightly to support the new textsecure.Content proto. I also added a new config option to allow users to turn off device syncing completely. 

Let me know what you think and if you'd like anything changed.

janimo commented:
Sorry I had no time to review this - I am pretty busy and distracted right now, but your changes are great, appreciated and will be merged one way or another :)
cheers nd good luck with the Sailfish client!

aebruno commented:
Thanks! Sounds great and no rush at all.

janimo commented on config.go:
```diff
@@ -11,17 +11,18 @@ import (
+	UserAgent             string `yaml:"userAgent"`             // Override for the default HTTP User Agent header field
+	AlwaysTrustPeerID     bool   `yaml:"alwaysTrustPeerID"`     // Workaround until proper handling of peer reregistering with new ID.
+	EnableMultiDeviceSync bool   `yaml:"enableMultiDeviceSync"` // Enable multi-device synchronization
 }
```
Any reason to ever disable this feature or did you set it while you considered it experimental?

aebruno replied:
Added this option because it was experimental but also can think of a few use cases where one might want to disable this. For example, your desktop was compromised and you want disable device syncing temporarily? or you deliberately want to ensure no SyncMessages are ever sent. This feature allows syncing of private data (contacts/groups) and perhaps one wants to completely disable this? You could always just unlink the device as well so not sure how useful it is to keep this option.

janimo replied:
Yeah, I can't think of compelling use cases for this option TBH. The user can control were to sync by managing the linked devices as you say, and not via a config option. If you personally hve no se for it now the code is not experimental, I'd say remove it, and have one less file touched by the changes :)

aebruno replied:
Sounds good, I agree. All removed.

janimo commented on cmd/textsecure/main.go:
```diff
@@ -221,6 +228,51 @@ func main() {
+		}
+
+		for _, d := range devs {
+			fmt.Printf("ID: %d\n", d.ID)
```
log instead of fmt here as well?

aebruno commented:
Added this option because it was experimental but also can think of a few use cases where one might want to disable this. For example,  your desktop was compromised and you want disable device syncing temporarily? or you deliberately want to ensure no SyncMessages are ever sent. This feature allows syncing of private data (contacts/groups) and perhaps one wants to completely disable this? You could always just unlink the device as well so not sure how useful it is to keep this option.

janimo commented on sync.go:
```diff
@@ -1,12 +1,207 @@
+	}
+
+	tmp, err := ioutil.TempFile(config.StorageDir, "multidevice-contact-update")
+	if err != nil {
```
Did you consider using a bytes.Buffer or a bytes.Reader instead of a tempfile in these two functions?

aebruno commented:
Sorry for the delay. Fixed up sync.go to use bytes.Buffer and changed fmt to log. Let me know if you spot anything else.
